proposals/manifest_trial_tokens.md

```diff
@@ -0,0 +1,176 @@
+   code
+ - Chromium explicitly introduced breaking Manifest Version transition (first
+   from Manifest Version 1 to Manifest Version 2, and now from Manifest Version
+   2 to Manifest Version 3). This is taxing on developers since developers may
+   have to maintain code paths for both API behaviors for the entire duration of
+   migration.
+
+The new manifest key `trial_tokens` can enable per-change and per-extension
+opt-in leading to speedup in API iteration.
+
+See also [Chromium source code](https://github.com/chromium/chromium/blob/main/chrome/common/extensions/api/_api_features.json)
+
+### Known Consumers
+
+None at the moment.
+
+## Specification
+
+### Schema
+
+This is a manifest change with no programmable API.
+
+This proposal extends manifest with optional `trial_tokens` member which
+contains an array of strings which represent the stringified trial tokens.
+
+Parsed and recognized tokens should be exposed via existing
+[`runtime.getManifest()`](https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/API/runtime/getManifest)
+API in `origin_trials`. Invalid or unrecognized tokens should be excluded
+from the returned manifest. This will enable extensions to detect recognized
+tokens and adjust accordingly.
+
+### Behavior
+
+#### Manifest parsing
+
+Upon manifest parsing, browser should parse `trial_tokens` like so:
+ - `trial_tokens` key must never cause installation errors, only benign
+   warnings
+ - browsers can limit the total number of parsed and/or accepted tokens
+ - browsers can limit the length of each token
+ - browsers can limit the total length of all parsed and/or accepted tokens
+ - browsers can validate token structure, content, and signature at parse time;
+   in particular, tokens should not contain control characters like new lines
+   and charaxter returns which could affect HTTP header serialization and
+   parsing
+
+#### Manifest parsing algorithm
+ 1. if `trial_tokens` is not present, return
+ 2. if `trial_tokens` is not an array or if `trial_tokens` is an empty array,
+    (optionally) log a benign warning and return
+ 3. for every element `token` of `trial_tokens`:
+    1. (optionally) if the maximum number of accepted tokens is reached,
+       skip all other tokens
+    2. if `token` is not a string, skip it and (optionally) log a benign
+       warning
+    3. if `token` is an empty string, skip it and (optionally) log a benign
+       warning
+    4. (optionally) if `token` is too long to be parsed, skip it and
+       (optionally) log a benign warning
+    5. if token has been encountered before, skip it and (optionally) log
+       a benign warning
+    6. (optionally) attempt to parse the token and validate it. If validation
+       fails, skip the token and (optionally) log a benign warning.
+       Note: browsers may disregard token expiration errors if they are not
+       sure about the correctness of system timestamp at the time of token
+       validation
```

Comment:
We don't necessarily need to include a note to this effect in the proposal, but it's worth noting that token validation MAY be an asynchronous process that requires exchanging information with a trial token management service.

Reply:
I added the following:
> Browsers may validate tokens asynchronously, if their implementation requires this.